src/network/network_client.cpp

```diff
@@ -1250,6 +1251,42 @@ void NetworkClientsToSpectators(CompanyID cid)
 	cur_company.Restore();
 }
 
+/**
+ * Check whether the given client name is deemed valid for use in network games.
+ * An empty name (null or '') is not valid as that is essentially no name at all.
+ * A name starting with white space is not valid for tab completion purposes.
+ * @param client_name The client name to check for validity.
+ * @return True iff the name is valid.
+ */
+bool NetworkIsValidClientName(const char *client_name)
+{
+	if (StrEmpty(client_name)) return false;
+	if (*client_name == ' ') return false;
+	return true;
+}
+
+/**
+ * Check whether the given client name is deemed valid (NetworkIsValidClientName),
+ * and when it is not valid show an error message telling the user to change the
+ * client name to something more suitable.
+ * This function is not suitable for ensuring a valid client name at the server.
+ * @return True iff the error message is being shown to the user.
+ */
+bool NetworkShowErrorMessageForInvalidClientName(char *client_name)
+{
+	if (!StrEmpty(client_name)) StrTrimInPlace(client_name);
```

Comment:
I have been looking at this for a good 10 minutes now. I get why this line is here, but it also feels weird.

Basically, this function only does validation, but sneaky also trims the string if needed. This means the client name is always validated properly and fixed if possible, but the timing seems a bit weird. At least it is not something I would have expected this function to do (it is also not documented it does :P).

So it made me think, can't we do this different. For the popup in my Network GUI, we can. As there we can show the error as soon as you changed the name that it is invalid, or trim it. For the console, we also trim it. Via settings we can do via a change-callback. That only leaves 2 places where it is more difficult:

- Startup with an invalid name.
- The main network GUI where the player name is just an edit box.

Startup should be easy to solve; pretty sure we have more code doing that. Maybe even the change-callback is called on load? I don't even know :D

The main network GUI is a bit more tricky. I would suggest either of the two solutions:
- Do like the new Network GUI, and only show the name, with an edit icon next to it to edit it. This gives a Query window, making it better to show the error after pressing OK
- Do the trimming after editing, but leave the error for later. That at least means this trimming can be removed from here, and put close to the places that manipulate it.

What do you think?
(sorry for the long story, but I didn't just want to say: I think this is wrong :D)

Reply:
> Startup should be easy to solve; pretty sure we have more code doing that. Maybe even the change-callback is called on load? I don't even know :D

I think we should not require that someone enters their name just to start the game. Only once you start using the network should it ask for a name.

> The main network GUI is a bit more tricky. I would suggest either of the two solutions:
> * Do like the new Network GUI, and only show the name, with an edit icon next to it to edit it. This gives a Query window, making it better to show the error after pressing OK
> * Do the trimming after editing, but leave the error for later. That at least means this trimming can be removed from here, and put close to the places that manipulate it.

The "fun" with the edit box is that it calls OnEditboxChanged after every key stroke. Now if you want to write your full name (given and surname) with a space in between, every time you type the space just before entering your last name it will be trimmed. That's going to cause a lot of bug reports, so sadly the trimming can only be done once we are fairly certain the user is done editing, which coincidentally is the moment they try to join a game or start a server.

Hiding the input field for the user name might actually make it harder for new users to find the place where to enter their name.